### PR TITLE
feat(ingestion): bigquery - Option to send usage queries as well as Operational metadata

### PR DESCRIPTION
**Pull request by treff7es**
## Checklist
- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

### CODE REVIEW

File: metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py [hunk 1]
```diff
@@ -462,10 +469,23 @@ def from_entry(cls, entry: AuditLogEntry) -> "QueryEvent":
             timestamp=entry.timestamp,
             actor_email=entry.payload["authenticationInfo"]["principalEmail"],
             query=job_query_conf["query"],
-        )
-        # jobName
-        query_event.jobName = _job_name_ref(
-            job.get("jobName", {}).get("projectId"), job.get("jobName", {}).get("jobId")
+            job_name=_job_name_ref(
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py:472]: curious why you went from more defensive style `x.get('attr',{})` to more fragile `x['attr']`. Is it guaranteed that we find these columns? What if permissions don't allow you to see these columns.

File: metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py [hunk 2]
```diff
@@ -474,9 +494,11 @@ def from_entry(cls, entry: AuditLogEntry) -> "QueryEvent":
                 raw_dest_table
             )
         # statementType
-        query_event.statementType = job_query_conf.get("statementType")
         # referencedTables
         job_stats: Dict = job["jobStatistics"]
+        if "totalBilledBytes" in job_stats and job_stats["totalBilledBytes"]:
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py:499]: can this be simplified to:
```python
if job_stats.get("totalBilledBytes"):
```

File: metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py (hunk 1 above)
Comment (treff7es) [reply metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py:472]: Thanks, you are right, I will fix it.

File: metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py [hunk 3]
```diff
@@ -521,22 +545,37 @@ def from_exported_bigquery_audit_metadata(
...
+            if job["jobStats"]["endTime"]
+            else None,
+            numAffectedRows=int(query_stats["outputRowCount"])
+            if "outputRowCount" in query_stats and query_stats["outputRowCount"]
+            else None,
+            statementType=query_config["statementType"],
         )
         # jobName
-        query_event.jobName = job.get("jobName")
+        query_event.job_name = job["jobName"]
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py:571]: `.get(` versus `[]` here as well

File: metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py [hunk 4]
```diff
@@ -548,17 +587,18 @@ def from_exported_bigquery_audit_metadata(
...
         query_event.payload = payload if DEBUG_INCLUDE_FULL_PAYLOADS else None
 
-        if not query_event.jobName:
+        if not query_event.job_name:
             logger.debug(
                 "jobName from query events is absent. "
                 "BigQueryAuditMetadata entry - {logEntry}".format(logEntry=row)
             )
 
+        if "totalBilledBytes" in query_stats and query_stats["totalBilledBytes"]:
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py:599]: can be replaced by `if query_stats.get("totalBilledBytes"):`

File: metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py [hunk 5]
```diff
@@ -598,15 +652,24 @@ def from_entry_v2(cls, row: BigQueryAuditMetadata) -> "QueryEvent":
...
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py:661]: can be replaced by `if query_stats.get("totalBilledBytes"):`

File: metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py [hunk 6]
```diff
@@ -855,13 +940,9 @@ def _get_bigquery_log_entries_via_gcp_logging(
...
         )
-        deny_regex = (
-            audit_templates["BQ_FILTER_REGEX_DENY_TEMPLATE"].format(
-                deny_pattern=self.config.get_deny_pattern_string(),
-                logical_operator="AND" if use_allow_filter else "",
-            )
-            if use_deny_filter
-            else ("" if use_allow_filter else "FALSE")
+        deny_regex = audit_templates["BQ_FILTER_REGEX_DENY_TEMPLATE"].format(
+            deny_pattern=f"__TABLES_SUMMARY__|INFORMATION_SCHEMA{'|'.join(self.config.get_deny_pattern_string())}",
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py:944]: `.join(string)` will treat string as a list, which is not what we want here.

File: metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py [hunk 7]
```diff
@@ -1106,77 +1275,77 @@ def event_processor(
...
             ):
                 continue
 
-            if event.jobName:
-                if event.jobName in query_jobs:
+            # There are some read event which does not have jobName because it was read in a different way
+            # Like https://cloud.google.com/logging/docs/reference/audit/bigquery/rest/Shared.Types/AuditData#tabledatalistrequest
+            # There are various reason to read a table
+            # https://cloud.google.com/bigquery/docs/reference/auditlogs/rest/Shared.Types/BigQueryAuditMetadata.TableDataRead.Reason
+            if event.read_event.jobName:
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py:1293]: `s/jobName/job_name/`

Comment (shirshanka) [reply metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py:1293]: ah looks like read_event has retained the old jobName field

File: metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py [hunk 8]
```diff
@@ -1106,77 +1275,77 @@ def event_processor(
...
+                if event.read_event.jobName in query_jobs:
                     # Join the query log event into the table read log event.
                     num_joined += 1
-                    event.query = query_jobs[event.jobName].query
-
-                    # TODO also join into the query itself for column references
+                    event.query_event = query_jobs[event.read_event.jobName]
+                    yield event
```
Comment (shirshanka) [inline metadata-ingestion/src/datahub/ingestion/source/usage/bigquery_usage.py:1298]: don't need this yield + continue as we are yielding finally anyway

**Review by shirshanka — APPROVED** [review]
LGTM

**Pull request merged**